torchaudio/csrc/sox/effects.cpp

```diff
@@ -137,51 +137,65 @@ c10::intrusive_ptr<TensorSignal> apply_effects_file(
     c10::optional<bool>& normalize,
     c10::optional<bool>& channels_first,
     c10::optional<std::string>& format) {
-  // Streaming decoding over file-like object is tricky because libsox operates
-  // on FILE pointer. The folloing is what `sox` and `play` commands do
-  //  - file input -> FILE pointer
-  //  - URL input -> call wget in suprocess and pipe the data -> FILE pointer
-  //  - stdin -> FILE pointer
+  // Prepare the buffer used throughout the lifecycle of SoxEffectChain.
   //
-  // We want to, instead, fetch byte strings chunk by chunk, consume them, and
-  // discard.
+  // For certain format (such as FLAC), libsox keeps reading the content at
+  // the initialization unless it reaches EOF even when the header is properly
+  // parsed. (Making buffer size 8192, which is way bigger than the header,
+  // resulted in libsox consuming all the buffer content at the time it opens
+  // the file.) Therefore buffer has to always contain valid data, except after
+  // EOF. We default to `sox_get_globals()->bufsiz`* for buffer size and we
+  // first check if there is enough data to fill the buffer. `read_fileobj`
+  // repeatedly calls `read`  method until it receives the requested lenght of
+  // bytes or it reaches EOF. If we get bytes shorter than requested, that means
+  // the whole audio data are fetched.
   //
-  // Here is the approach
-  // 1. Initialize sox_format_t using sox_open_mem_read, providing the initial
-  // chunk of byte string
-  //    This will perform header-based format detection, if necessary, then fill
-  //    the metadata of sox_format_t. Internally, sox_open_mem_read uses
-  //    fmemopen, which returns FILE* which points the buffer of the provided
-  //    byte string.
-  // 2. Each time sox reads a chunk from the FILE*, we update the underlying
-  // buffer in a way that it
-  //    starts with unseen data, and append the new data read from the given
-  //    fileobj. This will trick libsox as if it keeps reading from the FILE*
-  //    continuously.
-
-  // Prepare the buffer used throughout the lifecycle of SoxEffectChain.
-  // Using std::string and let it manage memory.
-  // 4096 is minimum size requried by auto_detect_format
-  // https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L40-L48
-  const size_t in_buffer_size = 4096;
-  std::string in_buffer(in_buffer_size, 'x');
-  auto* in_buf = const_cast<char*>(in_buffer.data());
-
-  // Fetch the header, and copy it to the buffer.
-  auto header = static_cast<std::string>(
-      static_cast<py::bytes>(fileobj.attr("read")(4096)));
-  memcpy(
-      static_cast<void*>(in_buf),
-      static_cast<void*>(const_cast<char*>(header.data())),
-      header.length());
+  // * This can be changed with `torchaudio.utils.sox_utils.set_buffer_size`.
+  auto capacity =
+      (sox_get_globals()->bufsiz > 256) ? sox_get_globals()->bufsiz : 256;
+  std::string buffer(capacity, '\0');
```

Comment:
Perhaps there is a better type on std instead of string to define your buffer? Perhaps `py::bytes`?

Reply:
Looking at [the implementation of `py::bytes`](https://github.com/pybind/pybind11/blob/48534089f7f1b0229bc7ae2e4d0f21dc9ad169b5/include/pybind11/pytypes.h#L1002-L1029), they do not provide direct access to the underlying memory. The only way to get an access to the memory is converting it to `std::string`. So I think it is straightforward to use `std::string`